lib/serverengine/winsock.rb

```diff
@@ -0,0 +1,193 @@
+      WSA_FLAG_NO_HANDLE_INHERIT = 0x80
+
+      # PROTOCOL
+      MAX_PROTOCOL_CHAIN = 7
+      WSAPROTOCOL_LENGTH = 256
+      GUID_DATA4_LENGTH = 8
+    end
+
+    include Constants
+
+    typedef :ulong, :dword
+    typedef :uintptr_t, :socket
+    typedef :pointer, :ptr
+    typedef :ushort, :word
+    typedef :uintptr_t, :handle
+
+    ffi_lib :ws2_32
+
+    attach_function :closesocket, [:socket], :int
+    attach_function :inet_addr, [:string], :ulong
+
+    attach_function :FreeAddrInfoEx, [:pointer], :void
+    attach_function :GetAddrInfo, :getaddrinfo, [:string, :string, :pointer, :pointer], :int
+    attach_function :GetAddrInfoW, [:buffer_in, :buffer_in, :pointer, :pointer], :int
+
+    attach_function :GetAddrInfoExA, [:string, :string, :dword, :ptr, :ptr, :ptr, :ptr, :ptr, :ptr, :ptr], :int
+    attach_function :GetHostByAddr, :gethostbyaddr, [:string, :int, :int], :pointer
+    attach_function :GetHostByName, :gethostbyname, [:string], :pointer
+    attach_function :GetProtoByName, :getprotobyname, [:string], :ptr
+    attach_function :GetProtoByNumber, :getprotobynumber, [:int], :ptr
+    attach_function :GetHostName, :gethostname, [:buffer_out, :int], :int
+    attach_function :GetServByName,:getservbyname, [:string, :string], :pointer
+
+    attach_function :WSAAsyncGetHostByAddr, [:uintptr_t, :uint, :string, :int, :int, :buffer_out, :int], :uintptr_t
+    attach_function :WSAAsyncGetHostByName, [:uintptr_t, :uint, :string, :buffer_out, :pointer], :uintptr_t
+    attach_function :WSAAsyncGetProtoByName, [:uintptr_t, :uint, :string, :buffer_out, :pointer], :uintptr_t
+    attach_function :WSAAsyncGetProtoByNumber, [:uintptr_t, :uint, :int, :buffer_out, :pointer], :uintptr_t
+    attach_function :WSAAsyncGetServByName, [:uintptr_t, :uint, :int, :string, :buffer_out, :int], :uintptr_t
+    attach_function :WSAAsyncGetServByPort, [:uintptr_t, :uint, :int, :string, :buffer_out, :int], :uintptr_t
+    attach_function :WSACancelAsyncRequest, [:uintptr_t], :int
+    attach_function :WSACleanup, [], :int
+    attach_function :WSAConnect, [:socket, :ptr, :int, :ptr, :ptr, :ptr, :ptr], :int
+    attach_function :WSAConnectByNameA, [:socket, :string, :string, :ptr, :ptr, :ptr, :ptr, :ptr, :ptr], :bool
+    attach_function :WSAEnumNameSpaceProvidersA, [:ptr, :ptr], :int
+    attach_function :WSAEnumProtocolsA, [:ptr, :ptr, :ptr], :int
+    attach_function :WSAGetLastError, [], :int
+    attach_function :WSASocketA, [:int, :int, :int, :ptr, :int, :dword], :socket
+    attach_function :WSAStartup, [:word, :ptr], :int
+    attach_function :WSADuplicateSocketA, [:socket, :dword, :ptr], :int
+    attach_function :bind, [:socket, :ptr, :int], :int
+    attach_function :listen, [:socket, :int], :int
+    attach_function :WSAAccept, [:socket, :ptr, :int, :int, :int], :socket
+    attach_function :accept, [:socket, :ptr, :int], :socket
+
+    rubydll_path = Dir.glob(RbConfig.expand("$(bindir)")+"/msvcrt-ruby*.dll").first
```

Comment:
This can adapt to ruby mswin32, mswin64 and mingw32. (msvcrt-ruby*.dll in mingw32 and mingw32, msvcr*-ruby*.dll in mswin64)
Cygwin is the different story here, because unix version run in cygwin.

I think it's enough because many users install ruby with Rubyinstaller or Visual Studio.
If it's wrong or there is better way, I'll fix it.